File: 2주차/박수교/정수 삼각형.py
def solution(triangle):
    #dp 테이블 구성
    dp = [[0] * i for i in range(1, len(triangle) + 1)]

    #초기값 설정
    dp[0][0] = triangle[0][0]
    dp[1][0] = triangle[1][0] + triangle[0][0]
    dp[1][1] = triangle[1][1]+ triangle[0][0]

    for i in range(2, len(triangle)):
        for j in range(len(triangle[i])):
            #인덱스 0번값 처리
            if j == 0:
                dp[i][j] = triangle[i][j] + dp[i-1][j]
                continue
            #인덱스 len(triangle[i])-1번 값 처리
            elif j == len(triangle[i])-1:
                dp[i][j] = triangle[i][j] + dp[i-1][j-1]
                continue
            #그 외 나머지 값들 처리
            dp[i][j] = triangle[i][j] + max(dp[i-1][j], dp[i-1][j-1])
    
    #마지막줄에서 가장 큰 값 return         
    return max(dp[len(triangle)-1])

File: 2주차/김태경/정수 삼각형.java
import java.util.*;
class Solution {
    int[][] dp;
    int[] dx={1,1};
    int[] dy={0,1};
    public int solution(int[][] triangle) {
        dp=new int[triangle.length][triangle[triangle.length-1].length];
        
        run(triangle);
        
        int answer = 0;
        for(int i=0;i<triangle[triangle.length-1].length;i++){
            //System.out.println(dp[triangle.length-1][i]);
            answer=Math.max(answer,dp[triangle.length-1][i]);
        }
        return answer;
    }
    
    void run(int[][] tri){
        dp[0][0]=tri[0][0];
        int i=0;
        int j=0;
        for(int u = 0;u<tri.length;u++)
        while (i<tri.length-1 && j<tri[tri.length-1].length-1){
            for(int k=0; k<2;k++){
                int nx=i+dx[k];
                int ny=j+dy[k];
                
                if(nx>=tri.length || ny>=tri[tri.length-1].length ){
                    continue;
                }
                dp[nx][ny]=Math.max(dp[nx][ny], dp[i][j]+tri[nx][ny]);
                
            }
            if(j==i){
                i++;
                j=0;
            }else{
                j++;
            }
        }
    }
}
